### PR TITLE
use subtest for table units (pkg/scheduler/util)

### LINKED ISSUE #63267: scheduler: Update tests to use subtest
Go 1.7 added the [subtest feature](https://golang.org/doc/go1.7#testing) which can make table-driven tests much easier to run and debug. Many table-driven tests in `pkg/scheduler` and `test/integration/scheduler` are not using this feature.

As an example, https://github.com/kubernetes/kubernetes/pull/56940 converts some tests to use a sub-benchmark.

Further reading: [Using Subtests and Sub-benchmarks](https://blog.golang.org/subtests)

/kind cleanup
/sig scheduling

### PR DESCRIPTION
**Pull request by xchapter7x**
**What this PR does / why we need it**: Update scheduler's unit table tests to use subtest

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

**Special notes for your reviewer**:
breaks up PR: https://github.com/kubernetes/kubernetes/pull/63281
/ref #63267

**Release note**:

```release-note
This PR will leverage subtests on the existing table tests for the scheduler units.
Some refactoring of error/status messages and functions to align with new approach.

```

### CODE REVIEW

File: pkg/scheduler/util/backoff_utils_test.go [hunk 1]
```diff
@@ -64,23 +64,33 @@ func TestBackoff(t *testing.T) {
...
-	if duration != 60*time.Second {
-		t.Errorf("expected: 60, got %s", duration.String())
-	}
-	// Verify that we split on namespaces correctly, same name, different namespace
-	fooID.Namespace = "other"
-	duration = backoff.GetEntry(fooID).getBackoff(backoff.maxDuration)
-	if duration != 1*time.Second {
-		t.Errorf("expected: 1, got %s", duration.String())
-	}
+	t.Run("Verify that we split on namespaces correctly, same name, different namespace", func(t *testing.T) {
```
Comment (misterikkit) [inline pkg/scheduler/util/backoff_utils_test.go]: I don't think this t.Run improves the clarity of the test. Similarly, the two subtests below might make more sense as better error messages.

File: pkg/scheduler/util/utils_test.go [hunk 2]
```diff
@@ -294,12 +300,14 @@ func TestHostPortInfo_Check(t *testing.T) {
...
 
 	for _, test := range tests {
-		hp := make(HostPortInfo)
-		for _, param := range test.added {
-			hp.Add(param.ip, param.protocol, param.port)
-		}
-		if hp.CheckConflict(test.check.ip, test.check.protocol, test.check.port) != test.expect {
-			t.Errorf("%v failed, expected %t; got %t", test.desc, test.expect, !test.expect)
-		}
+		t.Run(test.desc, func(t *testing.T) {
```
Comment (misterikkit) [inline pkg/scheduler/util/utils_test.go]: rename desc to name?

**Comment by k82cn** [comment]
@misterikkit , would you also help to review other PRs about subtest?

**Comment by misterikkit** [comment]
> @misterikkit , would you also help to review other PRs about subtest?

I think I replied to them all. Feel free to assign if I missed one. I think someone has to approve after I give LGTM though.

**Comment by k82cn** [comment]
great, I can help to approve them :)

File: pkg/scheduler/util/utils_test.go (hunk 2 above)
Comment (xchapter7x) [reply pkg/scheduler/util/utils_test.go]: 👍

File: pkg/scheduler/util/backoff_utils_test.go (hunk 1 above)
Comment (xchapter7x) [reply pkg/scheduler/util/backoff_utils_test.go]: agreed. i meant to come back and revert this bit, as its a bit out of scope for the add subtests on table driven units objective.

File: pkg/scheduler/util/backoff_utils_test.go [hunk 3]
```diff
@@ -64,13 +64,17 @@ func TestBackoff(t *testing.T) {
...
-		}
-		clock.t = clock.t.Add(test.advanceClock)
-		backoff.Gc()
+		t.Run(test.expectedDuration.String(), func(t *testing.T) {
+			duration := backoff.GetEntry(test.podID).getBackoff(backoff.maxDuration)
+			if duration != test.expectedDuration {
+				t.Errorf("expected: %s, got %s for %s", test.expectedDuration.String(), duration.String(), test.podID)
+			}
+
+			clock.t = clock.t.Add(test.advanceClock)
```
Comment (misterikkit) [inline pkg/scheduler/util/backoff_utils_test.go]: Hmmm... this is another test that expects the cases to be run in sequence and carries over state.

File: pkg/scheduler/util/utils_test.go [hunk 4]
```diff
@@ -294,12 +300,14 @@ func TestHostPortInfo_Check(t *testing.T) {
...
-		if hp.CheckConflict(test.check.ip, test.check.protocol, test.check.port) != test.expect {
-			t.Errorf("%v failed, expected %t; got %t", test.desc, test.expect, !test.expect)
-		}
+		t.Run(test.name, func(t *testing.T) {
+			hp := make(HostPortInfo)
+			for _, param := range test.added {
+				hp.Add(param.ip, param.protocol, param.port)
+			}
+			if hp.CheckConflict(test.check.ip, test.check.protocol, test.check.port) != test.expect {
+				t.Errorf("%v failed, expected %t; got %t", test.name, test.expect, !test.expect)
```
Comment (misterikkit) [inline pkg/scheduler/util/utils_test.go]: remove test.name from log

File: pkg/scheduler/util/utils_test.go [hunk 5]
```diff
@@ -190,69 +194,71 @@ func TestHostPortInfo_AddRemove(t *testing.T) {
...
-			hp.Add(param.ip, param.protocol, param.port)
-		}
-		for _, param := range test.removed {
-			hp.Remove(param.ip, param.protocol, param.port)
-		}
-		if hp.Len() != test.length {
-			t.Errorf("%v failed: expect length %d; got %d", test.desc, test.length, hp.Len())
-			t.Error(hp)
-		}
+		t.Run(test.desc, func(t *testing.T) {
```
Comment (misterikkit) [inline pkg/scheduler/util/utils_test.go]: please rename desc to name as you did with the other test.

File: pkg/scheduler/util/utils_test.go [hunk 6]
```diff
@@ -190,69 +194,71 @@ func TestHostPortInfo_AddRemove(t *testing.T) {
...
+		t.Run(test.name, func(t *testing.T) {
+			hp := make(HostPortInfo)
+			for _, param := range test.added {
+				hp.Add(param.ip, param.protocol, param.port)
+			}
+			for _, param := range test.removed {
+				hp.Remove(param.ip, param.protocol, param.port)
+			}
+			if hp.Len() != test.length {
+				t.Errorf("failed: expect length %d; got %d", test.length, hp.Len())
```
Comment (misterikkit) [inline pkg/scheduler/util/utils_test.go]: You might want to remove "failed:" from the message.

File: pkg/scheduler/util/utils_test.go [hunk 7]
```diff
@@ -294,12 +300,14 @@ func TestHostPortInfo_Check(t *testing.T) {
...
-		if hp.CheckConflict(test.check.ip, test.check.protocol, test.check.port) != test.expect {
-			t.Errorf("%v failed, expected %t; got %t", test.desc, test.expect, !test.expect)
-		}
+		t.Run(test.name, func(t *testing.T) {
+			hp := make(HostPortInfo)
+			for _, param := range test.added {
+				hp.Add(param.ip, param.protocol, param.port)
+			}
+			if hp.CheckConflict(test.check.ip, test.check.protocol, test.check.port) != test.expect {
+				t.Errorf("failed, expected %t; got %t", test.expect, !test.expect)
```
Comment (misterikkit) [inline pkg/scheduler/util/utils_test.go]: as above, please remove "failed,"

**Comment by xchapter7x** [comment]
@k82cn @misterikkit

All requested changes/feedback should now be addressed. Just waiting for the checks to go green.

-John

**Pull request closed without merging**